### ISSUE TITLE
Please provide help thx! r.keyboard() work in the wrong place - see this example

### ISSUE THREAD

**Issue by YuZhouQ** [tebelorg/RPA-Python#466]
r.init(visual_automation = True)
um.go_href('https://m.baidu.com/')
r.type('#index-form > div:nth-child(1) > div.con-wrap.new-search-con > div > div.input-wrapper > label','how much')
r.keyboard('[ctrl]a')
r.keyboard('[backspace]')

I want to ues keybord[backspace] to delete the words "how much" i type , but eventually it change the top search box.
![image](https://user-images.githubusercontent.com/101469027/233886987-a29b0e94-67fc-4d2b-8304-148919cd71fc.png)

**Comment by ck81** [comment]
~~~
r.type('#index-form > div:nth-child(1) > div.con-wrap.new-search-con > div > div.input-wrapper > label','how much')
~~~

Think it didn't work because the XPath odes not look right.  That one looks like CSS selector.

**Comment by ck81** [comment]
Just to add on...

1) For web-based applications, or automating anything in web pages, we usually try not to use visual automation, unless absolutely necessary.  I've been using TagUI for many years.  99.9% of the time, you can automate any web pages using just XPath.

2) The main reason for using just xpath is that: you can do other work while your TagUI script is running e.g. check your email in Outlook, work on your Powerpoint, or do calculations in Excel.  You can even lock your machine i.e. you start your TagUI robot running, lock your machine, go for your lunch, and when you're back, everything is completed!  The moment you used visual automation, similar to UiPath and most of the other RPA tools, it controls your mouse and keyboard, and you cannot do any other work while the robot is running.  And most important, you CANNOT lock your machine.

For your case, the script is as simple as the following:
~~~
import tagui as r
r.init()

r.url("https://m.baidu.com")
r.type('//input[contains(@id, "kw")]', '[clear]')

# the following just for testing
r.wait(1)
r.type('//input[contains(@id, "kw")]', 'RPA[enter]')
~~~

Note:

- Do not use `visual_automation = True` in `r.init()`
- Use `[clear]` to clear the input filed
- Your original xpath is not correct
- If you use the Chrome Developer Tool to view the HTML for the input field, it will show:
~~~
<input type="search" autocomplete="off" autocorrect="off" maxlength="64" id="index-kw" name="word" class="se-input adjust-input" data-sa="" placeholder="输入搜索词" data-sug="1">
~~~
The simplest xpath for the above is: `//input[@id="index-kw"]`

- However, I think baidu.com tries to stop people from using RPA on its website.  So every now and then, the code will change to the following:
~~~
<input autocomplete="off" autocorrect="off" maxlength="64" id="kw" name="word" type="search" class="se-input new-search-input" placeholder="输入搜索词" value="tagui" data-sug="1">
~~~

- Note that`id` has changed from `"index-kw"` to `"kw"`.  And the corresponding xpath will be: `//input[@id="kw"]`

- In order to handle both situations, you can use the following xpath : `//input[contains(@id, "kw")]`

- To enter search string into the field, you can use the code: `r.type('//input[contains(@id, "kw")]', 'RPA[enter]')`

Please give the above code a try and let me know if it works on your side.

One last note: How well you can automate a web page boils down to how good you are with XPath.  XPath is an industry standard to locate any elements on a webpage.  There are many tutorials out there on the Internet.  Just google for "xpath tutorial".  If you are not too familiar with XPath yet, I strongly recommend that you spend some time to build a strong foundation in XPath.

Once you have a good command of XPath, you will be able to read anything on a web page, type any text, click any button, or select any pulldown menu!

**Comment by kensoh** [comment]
Thanks @ck81 for sharing your solution! Adding on @YuZhouQ, for keyboard() function, you first need to use click() or type() with visual automation to actually make the focus on the text box. If you use click() or type() with XPath, the text box will appear to be active but it isn't actually where the key strokes will be sent if you try to type manually.

Something like:

```py
r.click('browser_icon.png') # to make browser active
r.click('text_box.png') # to set focus on textbox
r.keyboard('your message')
```

Alternatively, use XPath to avoid having to use keyboard() function and automate more easily. You can try out the SelectorsHub Chrome extension to read the XPath value easily.

**Comment by ck81** [comment]
@YuZhouQ replied in another thread to the above:

> Thank you very much for your detailed response to my question. And I'm sorry for my feedback being too late. As the issue was not resolved at the time, I changed my approach.
After reading your answer, I tried again using your method myself. The program ran very smoothly and successfully cleared after entering characters. More importantly, after reading your answer, I truly realized the beauty of xpath. And the differences between Tagui and other RPAs.
Meanwhile, what I want to ask is Is there any difference in the use of CSS_selector, xpath, and other element recognition methods? Is it necessary to keep the browser open when recognizing other elements besides xpath?
Thank you very much for your detailed answer, which has taught me a lot. I didn't have much understanding of xpath before, and it was only through exposure to rpa that I started to understand its use. I will increase my learning of it. 

> Meanwhile, what I want to ask is Is there any difference in the use of CSS_selector, xpath, and other element recognition methods?

XPath and CSS are the most popular industry standards.

### XPath vs CSS

TL;DR: If you can implement a query with CSS selectors, you can do the same with XPath. The other way round is not always possible.

* Xpath allows bidirectional flow which means the traversal can be both ways from parent to child and child to parent as well. CSS allows only one directional flow which means the traversal is from parent to child only.
* XPath expressions support multiple methods to find the elements which are difficult to locate due to dynamic attribute values or limited attributes. For example, if you need to locate an element that does not have any unique attribute, then you can use XPath axes methods (refer to below) to establish the parent-child relationship with another unique element. In this way, the required elements can be easily located.
* XPath has methods such as `Contains()`, `Starts-with()` which is very useful in scraping dynamic attribute value when you know some part of the attribute remains constant.
* Xpath allows identification with the help of visible text appearing on screen with the help of text() function. e.g. you can use `//b/text()` to identify the text content of all b elements. CSS does not have this feature. 

#### Note:XPath Axes Methods
XPath axes methods can traverse the DOM bidirectionally and are extremely useful in locating the elements that don’t have any attributes.

There are many axes methods, the most common ones are:
* `Following`
* `Preceding`
* `Ancestor`
* `Parent`
* `Child`

You can google for: `xpath vs css which is better`
And you will be able to find many articles that explain the above in details with examples.

>  Is it necessary to keep the browser open when recognizing other elements besides XPath?

No!  If you use 100% Xpath , then you can minimize the browser window and work on other applications such as Outlook or Excel when the TagUI runs.  (But I usually like to keep the browser window open in another monitor so that I can make sure that TagUi is running ok while I do my other work.)  You can even lock your machine and go for your lunch while TagUI runs.  Try this on other RPA tools.  More than 95% of the RPA tools out there in the market cannot do this.  That's why I like TagUI- truly impressive and powerful when it comes to automating web applications!

Again, let me emphasize one more time, this is only true if you use **100% Xpath**.  The moment you use visual automation with `r.init(visual_automation = True)`, you should not touch your keyboard and mouse while TagUI runs.
